
<div align="center">
    <img src="image/foto_portada.jpg" alt="Tres personas catando vinos." style="width: 600px; height: 300px; border-radius: 10px;">
    <h1 style="font-family: Arial, sans-serif; font-size: 32px; font-weight: bold; margin-top: 20px; text-align: center;">Explorando la calidad del vino</h1>
    <h2 style="font-family: Arial, sans-serif; font-size: 20px; font-weight: normal; text-align: center;">Una aproximación a través de técnicas de procesamiento de datos</h2>
</div>


<br>El objetivo de este trabajo es establecer, por medio de técnicas de procesamiento de datos, las variables más influyentes en la calidad sensorial del vino.<br> 
Se analiza una variable dependiente (calidad) y diez variables independientes: acidez volátil, ácido cítrico, azúcar residual, cloruros, dióxido de azufre libre, dióxido de azufre total, densidad, pH, sulfatos, y alcohol.<br>
Utilizaremos dos datasets enfocados en la calidad de vinos. Ambos se relacionan a las variantes de vino blanco y vino tinto del vino portugués "Vinho Verde".
Los conjuntos de datos utilizados en este proyecto son provenientes del Machine Learning Repository de UCI. Si deseas obtener más información acerca de estos conjuntos de datos, puedes consultar el siguiente enlace: [UCI Wine Quality Dataset](https://archive.ics.uci.edu/dataset/186/wine+quality).

 # Data  wrangling

## Importacion de datos

In [1]:
#librerias..
import pandas as pd

In [2]:
#tablas a importar
tablas = ['winequality-red.csv',
          'winequality-white.csv']

archivos = []

ruta_data = 'data'  # Ruta relativa a la carpeta "data"

#Importar los archivos CSV y guardarlos en una lista
for t in tablas:
    ruta_archivo = f'{ruta_data}/{t}'
    datos = pd.read_csv(ruta_archivo,
                        sep = ';',
                        encoding='latin-1',
                        low_memory=False)
    archivos.append(datos)

#Desempaquetado de archivos importados
vino_tinto,vino_blanco = archivos[:2]


## Descubrimiento de datos

In [3]:
# Cuántas filas y columnas tienen los conjuntos de datos
print("Información sobre las filas y columnas de los conjuntos de datos:")
print(f"--> Vino tinto\nFilas: {vino_tinto.shape[0]}\nColumnas: {vino_tinto.shape[1]}")
print("----------------------")
print(f"--> Vino blanco\nFilas: {vino_blanco.shape[0]}\nColumnas: {vino_blanco.shape[1]}")

Información sobre las filas y columnas de los conjuntos de datos:
--> Vino tinto
Filas: 1599
Columnas: 12
----------------------
--> Vino blanco
Filas: 4898
Columnas: 12


In [4]:
#primeros diez registros data set tinto
vino_tinto.head(10)

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   
5            7.4              0.66         0.00             1.8      0.075   
6            7.9              0.60         0.06             1.6      0.069   
7            7.3              0.65         0.00             1.2      0.065   
8            7.8              0.58         0.02             2.0      0.073   
9            7.5              0.50         0.36             6.1      0.071   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   
5                 13.0                  40.0   0.9978  3.51       0.56   
6                 15.0                  59.0   0.9964  3.30       0.46   
7                 15.0                  21.0   0.9946  3.39       0.47   
8                  9.0                  18.0   0.9968  3.36       0.57   
9                 17.0                 102.0   0.9978  3.35       0.80   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5  
5      9.4        5  
6      9.4        5  
7     10.0        7  
8      9.5        7  
9     10.5        5

In [5]:
#primeros diez registros data set blanco
vino_blanco.head(10)

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.0              0.27         0.36            20.7      0.045   
1            6.3              0.30         0.34             1.6      0.049   
2            8.1              0.28         0.40             6.9      0.050   
3            7.2              0.23         0.32             8.5      0.058   
4            7.2              0.23         0.32             8.5      0.058   
5            8.1              0.28         0.40             6.9      0.050   
6            6.2              0.32         0.16             7.0      0.045   
7            7.0              0.27         0.36            20.7      0.045   
8            6.3              0.30         0.34             1.6      0.049   
9            8.1              0.22         0.43             1.5      0.044   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 45.0                 170.0   1.0010  3.00       0.45   
1                 14.0                 132.0   0.9940  3.30       0.49   
2                 30.0                  97.0   0.9951  3.26       0.44   
3                 47.0                 186.0   0.9956  3.19       0.40   
4                 47.0                 186.0   0.9956  3.19       0.40   
5                 30.0                  97.0   0.9951  3.26       0.44   
6                 30.0                 136.0   0.9949  3.18       0.47   
7                 45.0                 170.0   1.0010  3.00       0.45   
8                 14.0                 132.0   0.9940  3.30       0.49   
9                 28.0                 129.0   0.9938  3.22       0.45   

   alcohol  quality  
0      8.8        6  
1      9.5        6  
2     10.1        6  
3      9.9        6  
4      9.9        6  
5     10.1        6  
6      9.6        6  
7      8.8        6  
8      9.5        6  
9     11.0        6

In [6]:
# Resumen del dataset de vino tinto
vino_tinto.info()
vino_tinto.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1599 non-null   float64
 1   volatile acidity      1599 non-null   float64
 2   citric acid           1599 non-null   float64
 3   residual sugar        1599 non-null   float64
 4   chlorides             1599 non-null   float64
 5   free sulfur dioxide   1599 non-null   float64
 6   total sulfur dioxide  1599 non-null   float64
 7   density               1599 non-null   float64
 8   pH                    1599 non-null   float64
 9   sulphates             1599 non-null   float64
 10  alcohol               1599 non-null   float64
 11  quality               1599 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


fixed acidity  volatile acidity  citric acid  residual sugar  \
count    1599.000000       1599.000000  1599.000000     1599.000000   
mean        8.319637          0.527821     0.270976        2.538806   
std         1.741096          0.179060     0.194801        1.409928   
min         4.600000          0.120000     0.000000        0.900000   
25%         7.100000          0.390000     0.090000        1.900000   
50%         7.900000          0.520000     0.260000        2.200000   
75%         9.200000          0.640000     0.420000        2.600000   
max        15.900000          1.580000     1.000000       15.500000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  1599.000000          1599.000000           1599.000000  1599.000000   
mean      0.087467            15.874922             46.467792     0.996747   
std       0.047065            10.460157             32.895324     0.001887   
min       0.012000             1.000000              6.000000     0.990070   
25%       0.070000             7.000000             22.000000     0.995600   
50%       0.079000            14.000000             38.000000     0.996750   
75%       0.090000            21.000000             62.000000     0.997835   
max       0.611000            72.000000            289.000000     1.003690   

                pH    sulphates      alcohol      quality  
count  1599.000000  1599.000000  1599.000000  1599.000000  
mean      3.311113     0.658149    10.422983     5.636023  
std       0.154386     0.169507     1.065668     0.807569  
min       2.740000     0.330000     8.400000     3.000000  
25%       3.210000     0.550000     9.500000     5.000000  
50%       3.310000     0.620000    10.200000     6.000000  
75%       3.400000     0.730000    11.100000     6.000000  
max       4.010000     2.000000    14.900000     8.000000

In [7]:
# Resumen del dataset de vino blanco
vino_blanco.info()
vino_blanco.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4898 entries, 0 to 4897
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         4898 non-null   float64
 1   volatile acidity      4898 non-null   float64
 2   citric acid           4898 non-null   float64
 3   residual sugar        4898 non-null   float64
 4   chlorides             4898 non-null   float64
 5   free sulfur dioxide   4898 non-null   float64
 6   total sulfur dioxide  4898 non-null   float64
 7   density               4898 non-null   float64
 8   pH                    4898 non-null   float64
 9   sulphates             4898 non-null   float64
 10  alcohol               4898 non-null   float64
 11  quality               4898 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 459.3 KB


fixed acidity  volatile acidity  citric acid  residual sugar  \
count    4898.000000       4898.000000  4898.000000     4898.000000   
mean        6.854788          0.278241     0.334192        6.391415   
std         0.843868          0.100795     0.121020        5.072058   
min         3.800000          0.080000     0.000000        0.600000   
25%         6.300000          0.210000     0.270000        1.700000   
50%         6.800000          0.260000     0.320000        5.200000   
75%         7.300000          0.320000     0.390000        9.900000   
max        14.200000          1.100000     1.660000       65.800000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  4898.000000          4898.000000           4898.000000  4898.000000   
mean      0.045772            35.308085            138.360657     0.994027   
std       0.021848            17.007137             42.498065     0.002991   
min       0.009000             2.000000              9.000000     0.987110   
25%       0.036000            23.000000            108.000000     0.991723   
50%       0.043000            34.000000            134.000000     0.993740   
75%       0.050000            46.000000            167.000000     0.996100   
max       0.346000           289.000000            440.000000     1.038980   

                pH    sulphates      alcohol      quality  
count  4898.000000  4898.000000  4898.000000  4898.000000  
mean      3.188267     0.489847    10.514267     5.877909  
std       0.151001     0.114126     1.230621     0.885639  
min       2.720000     0.220000     8.000000     3.000000  
25%       3.090000     0.410000     9.500000     5.000000  
50%       3.180000     0.470000    10.400000     6.000000  
75%       3.280000     0.550000    11.400000     6.000000  
max       3.820000     1.080000    14.200000     9.000000

## Eliminación de duplicados

In [8]:
# Función para verificar y eliminar duplicados en un dataset
def verificar_duplicados(dataset_nombre, dataset):
    print(f"Análisis de duplicados en el conjunto de datos {dataset_nombre.upper()}:")
   
    duplicados = dataset.duplicated().any()
    print(f"¿Hay duplicados? {duplicados}")
    
    cantidad_duplicados = dataset.duplicated().sum()
    print(f"Cantidad de duplicados: {cantidad_duplicados}")
    
    dataset = dataset.drop_duplicates()
    
    duplicados_despues = dataset.duplicated().any()
    print(f"¿Hay duplicados después de eliminarlos? {duplicados_despues}")
    
    print()  # Salto de línea
    
    return dataset

# Uso de la función para el conjunto de vino tinto
vino_tinto = verificar_duplicados("vino tinto", vino_tinto)

# Uso de la función para el conjunto de vino blanco
vino_blanco = verificar_duplicados("vino blanco", vino_blanco)

Análisis de duplicados en el conjunto de datos VINO TINTO:
¿Hay duplicados? True
Cantidad de duplicados: 240
¿Hay duplicados después de eliminarlos? False

Análisis de duplicados en el conjunto de datos VINO BLANCO:
¿Hay duplicados? True
Cantidad de duplicados: 937
¿Hay duplicados después de eliminarlos? False



## Transformacion y consolidacion de los datos

In [9]:
# Concatenación de los conjuntos de datos de vino tinto y vino blanco EN DF_VINOS
df_vinos = pd.concat([vino_tinto, vino_blanco], axis=0, ignore_index=True)  # Concatenación por filas

#validacion d ela integracion

def validar_filas(filas_esperadas, df_filas):
    if df_filas == filas_esperadas:
        print("La validación de filas fue exitosa")
    else:
        print("La validación de filas no se realizó correctamente")
        print(f"Filas esperadas: {filas_esperadas}")
        print(f"Filas encontradas: {df_filas}")


def validar_columnas(columnas_esperadas, columnas_df):
    columnas_faltantes = set(columnas_esperadas) - set(columnas_df)

    if columnas_faltantes:
        print("\nLa validación de columnas no se realizó correctamente. Faltan las siguientes columnas:")
        for columna in columnas_faltantes:
            print(columna)
    else:
        print("\nLa validación de columnas fue exitosa. Todas las columnas esperadas están presentes en el DataFrame resultante.")


def validar_integracion(dataset1, dataset2, dataset_integrado):
    # Validación de filas
    filas_esperadas = dataset1.shape[0] + dataset2.shape[0]
    dataset_integrado_filas = dataset_integrado.shape[0]
    validar_filas(filas_esperadas, dataset_integrado_filas)

    # Validación de columnas
    columnas_esperadas = vino_tinto.columns
    dataset_integrado_columnas = dataset_integrado.columns
    validar_columnas(columnas_esperadas, dataset_integrado_columnas)
    
validar_integracion(vino_tinto, vino_blanco, df_vinos)


La validación de filas fue exitosa

La validación de columnas fue exitosa. Todas las columnas esperadas están presentes en el DataFrame resultante.


In [10]:
# Diccionario de nuevos nombres de columnas
nuevos_nombres_col = {
    "fixed acidity": "acidez_fija",
    "volatile acidity": "acidez_volatil",
    "citric acid": "acido_citrico",
    "residual sugar": "azucar_residual",
    "chlorides": "cloruros",
    "free sulfur dioxide": "dioxido_azufre_libre",
    "total sulfur dioxide": "dioxido_azufre_total",
    "density": "densidad",
    "sulphates": "sulfatos",
    "quality": "calidad"
}

# Renombrar columnas del DataFrame df_vinos
df_vinos = df_vinos.rename(columns=nuevos_nombres_col)

In [11]:
#informacion del data set integrado
df_vinos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5320 entries, 0 to 5319
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   acidez_fija           5320 non-null   float64
 1   acidez_volatil        5320 non-null   float64
 2   acido_citrico         5320 non-null   float64
 3   azucar_residual       5320 non-null   float64
 4   cloruros              5320 non-null   float64
 5   dioxido_azufre_libre  5320 non-null   float64
 6   dioxido_azufre_total  5320 non-null   float64
 7   densidad              5320 non-null   float64
 8   pH                    5320 non-null   float64
 9   sulfatos              5320 non-null   float64
 10  alcohol               5320 non-null   float64
 11  calidad               5320 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 498.9 KB


**Observaciones Finales:**
<p align="left">El dataset consta de 5320 filas y 12 columnas. Todos los valores en las columnas tienen conteo no nulo. Los tipos de datos de las columnas son principalmente float64, excepto la columna "calidad" que es de tipo int64. El tamaño ocupado en memoria es aproximadamente 498.9 KB.</p>

## almacenamiento data set

In [12]:
# Guardar el dataset final en un archivo CSV
ruta_archivo_final = 'data/df_vinos.csv'  
df_vinos.to_csv(ruta_archivo_final, index=False)


In [13]:
# Imprimir mensaje de finalización del data wrangling
print("Proceso de data wrangling completado con éxito.")
print("El dataset ha sido preparado y está listo para su análisis.")

Proceso de data wrangling completado con éxito.
El dataset ha sido preparado y está listo para su análisis.
